### PR TITLE
Support runtime gltf operations.

### PR DESCRIPTION
**Pull request by fire**
For review.

How to use?

```C++
var gstate : GLTFState = GLTFState.new()
var gltf : PackedSceneGLTF = PackedSceneGLTF.new()
print(path);
var raw_f = File.new()
if raw_f.open(path, File.READ) != OK:
	return FAILED
var buffer : PackedByteArray = raw_f.get_buffer(raw_f.get_len())
var root_node : Node = gltf.import_gltf_scene_buffer(buffer, flags, 1000.0, gstate)
```

### CODE REVIEW

File: modules/gltf/editor_scene_importer_gltf.cpp [hunk 1]
```diff
@@ -78,13 +79,47 @@ void PackedSceneGLTF::_bind_methods() {
 			&PackedSceneGLTF::pack_gltf, DEFVAL(0), DEFVAL(1000.0f), DEFVAL(Ref<GLTFState>()));
 	ClassDB::bind_method(D_METHOD("import_gltf_scene", "path", "flags", "bake_fps", "state"),
 			&PackedSceneGLTF::import_gltf_scene, DEFVAL(0), DEFVAL(1000.0f), DEFVAL(Ref<GLTFState>()));
+	ClassDB::bind_method(D_METHOD("import_gltf_scene_buffer", "buffers", "flags", "bake_fps", "state"),
```
Comment (lyuma) [inline modules/gltf/editor_scene_importer_gltf.cpp]: "buffer", not "buffers"

File: modules/gltf/editor_scene_importer_gltf.cpp [hunk 2]
```diff
@@ -78,13 +79,47 @@ void PackedSceneGLTF::_bind_methods() {
...
+	if (r_state->animations.size()) {
+		AnimationPlayer *ap = memnew(AnimationPlayer);
+		root->add_child(ap);
+		ap->set_owner(root);
+		for (int i = 0; i < r_state->animations.size(); i++) {
+			gltf_document->_import_animation(r_state, ap, i, p_bake_fps);
+		}
+	}
+
+	return cast_to<Node3D>(root);
```
Comment (lyuma) [inline modules/gltf/editor_scene_importer_gltf.cpp]: This block of code should be moved into a function to be shared with the non _buffer version

File: modules/gltf/gltf_document.cpp [hunk 3]
```diff
@@ -6468,6 +6468,162 @@ Error GLTFDocument::parse(Ref<GLTFState> state, String p_path, bool p_read_binar
...
+	if (err != OK)
+		return Error::FAILED;
+
+	/* STEP 1 PARSE NODES */
+	err = _parse_nodes(state);
+	if (err != OK)
+		return Error::FAILED;
+
+	/* STEP 2 PARSE BUFFERS */
+	err = _parse_buffers(state, "");
```
Comment (lyuma) [inline modules/gltf/gltf_document.cpp:6615]: _parse_buffers must be modified to guarantee that no URI resolves are done if p_base_path is ""

File: modules/gltf/gltf_document.cpp [hunk 4]
```diff
@@ -6468,6 +6468,162 @@ Error GLTFDocument::parse(Ref<GLTFState> state, String p_path, bool p_read_binar
...
+	if (err != OK)
+		return Error::FAILED;
+
+	/* STEP 4 PARSE ACCESSORS */
+	err = _parse_accessors(state);
+	if (err != OK)
+		return Error::FAILED;
+
+	/* STEP 5 PARSE IMAGES */
+	err = _parse_images(state, "");
```
Comment (lyuma) [inline modules/gltf/gltf_document.cpp:6630]: _parse_buffers must be modified to guarantee that no URI resolves are done if p_base_path is ""

File: modules/gltf/gltf_document.cpp [hunk 5]
```diff
@@ -6468,6 +6468,162 @@ Error GLTFDocument::parse(Ref<GLTFState> state, String p_path, bool p_read_binar
...
+
+	ERR_FAIL_COND_V(!state->json.has("asset"), Error::FAILED);
+
+	Dictionary asset = state->json["asset"];
+
+	ERR_FAIL_COND_V(!asset.has("version"), Error::FAILED);
+
+	String version = asset["version"];
+
+	state->major_version = version.get_slice(".", 0).to_int();
```
Comment (lyuma) [inline modules/gltf/gltf_document.cpp:6601]: Everything from this point on must be merged into a common function with `parse()` and pass "" as the p_base_path, which we will handle specially later on.

**Review by lyuma — CHANGES_REQUESTED** [review]
Mostly good. We should make sure to add explicit security checks and migrate any URI resolving code for textures & buffers & etc. into a common helper which has explicit "" check that returns ERR_FAIL_...

**Comment by fire** [comment]
PackedScene should have a convert function from EditorSceneImporterMeshNode to MeshInstance

Edited: Done.

**Comment by fire** [comment]
> @fire it looks good. We got quite a lot of demand to make the GLTF importer part of core, given it is needed for VR and other stuff, so I have the feeling we should optimize it for performance (which we should do anyway) and have an option to load actual meshes instead of EditorImporterMesh.. so that means we would need to provide a function to parse the mesh node into a scene node and the editor supplies, otherwise it uses Mesh as a default

From reduz

**Pull request closed without merging**